File: test/brfc/expr/BinaryOperator_test.cpp
/*
Copyright 2010 Estonian Meteorological and Hydrological Institute

This file is part of baltrad-db.

baltrad-db is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

baltrad-db is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with baltrad-db. If not, see <http://www.gnu.org/licenses/>.
*/

#include <brfc/expr/BinaryOperator.hpp>
#include <brfc/expr/MockExpression.hpp>

namespace brfc {
namespace expr {

using ::testing::Ref;
using ::testing::Return;

class expr_BinaryOperator_test : public ::testing::Test {
  public:
    expr_BinaryOperator_test()
            : xpr1()
            , xpr2()
            , xpr1p(&xpr1, no_delete)
            , xpr2p(&xpr2, no_delete) {

    }

    virtual void SetUp() {
        ON_CALL(xpr1, clone())
            .WillByDefault(Return(xpr1p));
        ON_CALL(xpr1, do_equals(Ref(xpr1)))
            .WillByDefault(Return(true));
        ON_CALL(xpr2, clone())
            .WillByDefault(Return(xpr2p));
        ON_CALL(xpr2, do_equals(Ref(xpr2)))
            .WillByDefault(Return(true));
    }

    ::testing::NiceMock<MockExpression> xpr1;
    ::testing::NiceMock<MockExpression> xpr2;
    ExpressionPtr xpr1p, xpr2p;
};

TEST_F(expr_BinaryOperator_test, test_equals) {
    BinaryOperator op1(BinaryOperator::ADD, xpr1, xpr2);
    BinaryOperator op2(BinaryOperator::ADD, xpr2, xpr1);
    BinaryOperator op3(BinaryOperator::MUL, xpr1, xpr2);

    EXPECT_TRUE(op1.equals(op1));
    EXPECT_TRUE(op1.equals(op2));
    EXPECT_FALSE(op1.equals(op3));
    EXPECT_TRUE(op2.equals(op1));
    EXPECT_TRUE(op2.equals(op2));
    EXPECT_FALSE(op2.equals(op3));
    EXPECT_FALSE(op3.equals(op1));
    EXPECT_FALSE(op3.equals(op2));
    EXPECT_TRUE(op3.equals(op3));
}

} // namespace expr
} // namespace brfc

File: client/java/test/eu/baltrad/bdb/expr/proc/BinaryOperatorTest.java
/*
Copyright 2010-2011 Estonian Meteorological and Hydrological Institute

This file is part of baltrad-db.

baltrad-db is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

baltrad-db is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with baltrad-db. If not, see <http://www.gnu.org/licenses/>.
*/

package eu.baltrad.bdb.expr.proc;

import eu.baltrad.bdb.expr.Expression;
import eu.baltrad.bdb.expr.ExpressionFactory;
import eu.baltrad.bdb.expr.ListExpression;

import static org.junit.Assert.*;
import org.junit.Before;
import org.junit.Test;

import static org.easymock.EasyMock.*;
import org.easymock.EasyMockSupport;

public class BinaryOperatorTest extends EasyMockSupport {
  BinaryOperator classUnderTestMock;
  BinaryOperator classUnderTest;

  @Before
  public void setUp() {
    classUnderTestMock = createMock(BinaryOperator.class);
    classUnderTest = new BinaryOperator() {
      @Override
      protected Expression executeBinary(Expression lhs, Expression rhs) {
        return classUnderTestMock.executeBinary(lhs, rhs);
      }
    };
    resetAll();
  }

  @Test
  public void execute() {
    Expression lhs = createMock(Expression.class);
    Expression rhs = createMock(Expression.class);
    Expression expected = createMock(Expression.class);

    ListExpression args = new ListExpression();
    args.add(lhs);
    args.add(rhs);

    expect(classUnderTestMock.executeBinary(lhs, rhs))
      .andReturn(expected);
    replayAll();

    Expression result = classUnderTest.execute(args);
    assertSame(expected, result);

    verifyAll();
  }

  @Test
  public void execute_tooFewArguments() {
    ListExpression args = new ListExpression();
    try {
      classUnderTest.execute(args);
      fail("expected EvaluationError");
    } catch (EvaluationError e) { }
  }

  @Test
  public void execute_tooManyArguments() {
    ListExpression args = new ListExpression();
    args.add(createMock(Expression.class));
    args.add(createMock(Expression.class));
    args.add(createMock(Expression.class));
    try {
      classUnderTest.execute(args);
      fail("expected EvaluationError");
    } catch (EvaluationError e) { }
  }
}
